src/ir/possible-contents.cpp

```diff
@@ -0,0 +1,1628 @@
+    addRoot(curr);
+  }
+  void visitRefFunc(RefFunc* curr) {
+    addRoot(curr, PossibleContents::literal(Literal(curr->func, curr->type)));
+  }
+  void visitRefEq(RefEq* curr) {
+    // TODO: optimize when possible (e.g. when both sides must contain the same
+    //       global)
+    addRoot(curr);
+  }
+  void visitTableGet(TableGet* curr) {
+    // TODO: optimize when possible
+    addRoot(curr);
+  }
+  void visitTableSet(TableSet* curr) {}
+  void visitTableSize(TableSize* curr) { addRoot(curr); }
+  void visitTableGrow(TableGrow* curr) { addRoot(curr); }
+
+  void visitNop(Nop* curr) {}
+  void visitUnreachable(Unreachable* curr) {}
+
+#ifndef NDEBUG
+  // For now we only handle pops in a catch body, see visitTry(). To check for
+  // errors, use counter of the pops we handled and all the pops; those sums
+  // must agree at the end, or else we've seen something we can't handle.
+  Index totalPops = 0;
+  Index handledPops = 0;
+#endif
+
+  void visitPop(Pop* curr) {
+#ifndef NDEBUG
+    totalPops++;
+#endif
+  }
+  void visitI31New(I31New* curr) {
+    // TODO: optimize like struct references
+    addRoot(curr);
+  }
+  void visitI31Get(I31Get* curr) {
+    // TODO: optimize like struct references
+    addRoot(curr);
+  }
+
+  void visitRefTest(RefTest* curr) {
+    // TODO: optimize when possible
+    addRoot(curr);
+  }
+  void visitRefCast(RefCast* curr) {
+    // We will handle this in a special way later during the flow, as ref.cast
+    // only allows valid values to flow through.
+    addSpecialChildParentLink(curr->ref, curr);
+  }
+  void visitBrOn(BrOn* curr) {
+    // TODO: optimize when possible
+    handleBreakValue(curr);
+    receiveChildValue(curr->ref, curr);
+  }
+  void visitRttCanon(RttCanon* curr) { addRoot(curr); }
+  void visitRttSub(RttSub* curr) { addRoot(curr); }
+  void visitRefAs(RefAs* curr) {
+    // TODO optimize when possible: like RefCast, not all values flow through.
+    receiveChildValue(curr->value, curr);
+  }
+
+  // Locals read and write to their index.
+  // TODO: we could use a LocalGraph for SSA-like precision
+  void visitLocalGet(LocalGet* curr) {
+    if (isRelevant(curr->type)) {
+      for (Index i = 0; i < curr->type.size(); i++) {
+        info.links.push_back({LocalLocation{getFunction(), curr->index, i},
+                              ExpressionLocation{curr, i}});
+      }
+    }
+  }
+  void visitLocalSet(LocalSet* curr) {
+    if (!isRelevant(curr->value->type)) {
+      return;
+    }
+    for (Index i = 0; i < curr->value->type.size(); i++) {
+      info.links.push_back({ExpressionLocation{curr->value, i},
+                            LocalLocation{getFunction(), curr->index, i}});
+
+      // Tees also flow out the value (receiveChildValue will see if this is a
+      // tee based on the type, automatically).
+      receiveChildValue(curr->value, curr);
+    }
+  }
+
+  // Globals read and write from their location.
+  void visitGlobalGet(GlobalGet* curr) {
```

Comment:
Oh, do we allow tuples in globals? I didn't know that...

I'll add a TODO here. Fixing it will require adding a tupleIndex to GlobalLocation, I think, and maybe more. Probably not urgent as I don't think any code exists that uses tuple globals though.